### PR TITLE
Adds iterator and client unit tests, and prepares for more fetch failure handling

### PR DESCRIPTION
**Pull request by abellina**
This is an initial PR introducing the `RapidsShuffleFetchFailedException` and some usage, to try and address some of the issues brought up in: https://github.com/NVIDIA/spark-rapids/issues/326

It also adds initial unit tests for the client side.

Note that the write side has a change squeezed in there. When a task fails, it will call the writer with `stop(false)` to signify a failure during its processing. Prior to this PR, we would shut down the storage all-together => this would cause all peers to *this* executor to fail in the future, given that we would now reuse the executor's JVM. This change tracks these "uncomitted" ShuffleBufferIds at the writer level. @jlowe FYI, happy to split that off, handle it differently, but adding it here for discussion.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala [hunk 1]
```diff
@@ -501,18 +558,15 @@ class RapidsShuffleClient(
...
     */
-  private[this] def doHandleMetadataResponse(tx: Transaction,
-                                             resp: RefCountedDirectByteBuffer,
-                                             shuffleRequests: Seq[ShuffleBlockBatchId],
-                                             handler: RapidsShuffleFetchHandler): Unit = {
+  private[this] def doHandleMetadataResponse(
+      tx: Transaction,
+      resp: RefCountedDirectByteBuffer,
+      shuffleRequests: Seq[ShuffleBlockBatchId],
+      handler: RapidsShuffleFetchHandler): Unit = GpuShuffleEnv.synchronized{
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:565]: Why do we need to synchronize here?

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:565]: whoops. I had some non-thread-safe counters I was using to test this in a real job, and forgot to remove this.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleBufferCatalog.scala [hunk 2]
```diff
@@ -201,6 +201,15 @@ class ShuffleBufferCatalog(
...
     acquireBuffer(shuffleBufferId)
   }
+
+  /**
+    * Remove a buffer and table given a buffer ID
+    * @param id buffer identifier
+    */
+  def removeBuffer(id: ShuffleBufferId): Unit = {
+    tableMap.remove(id.tableId)
+    catalog.removeBuffer(id)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleBufferCatalog.scala:211]: Without any synchronization while modifying two related concurrent maps, is the assumption nobody should be asking for this shuffle buffer ID at the point this is called?  Elsewhere in the code that does this there's a comment explaining why there's no synchronization, wondering if we either have a sync issue here or should have a similar comment explaining why sync is unnecessary.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala [hunk 3]
```diff
@@ -33,8 +33,27 @@ import org.apache.spark.storage.ShuffleBlockBatchId
...
 trait RapidsShuffleFetchHandler {
+  /**
+    * After a response for shuffle metadata is received, the expected number of columnar
+    * batches is communicated back to the caller via this method.
+    * @param expectedBatches - number of columnar batches
+    */
   def start(expectedBatches: Int): Unit
+
+  /**
+    * Called when a buffer is received, and has been handed off to the catalog.
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:44]:
```suggestion
    * Called when a buffer is received and has been handed off to the catalog.
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala [hunk 4]
```diff
@@ -33,8 +33,27 @@ import org.apache.spark.storage.ShuffleBlockBatchId
...
+    * @param bufferId - a tracked shuffle buffer id
+    */
   def batchReceived(bufferId: ShuffleReceivedBufferId): Unit
+
+  /**
+    * Called when an error is detected during a metadata or a buffer fetch.
+    *
+    * @note No retry mechanism currently exists before calling [[transferError]],
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:52]: This note seems like an implementation detail of the code using this interface and not so relevant to those implementing the interface.  Does this make sense to place here?  Will we remember to update the comment when that is no longer the case, given the code that will be updated isn't here?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala]: Removed in https://github.com/NVIDIA/spark-rapids/pull/331/commits/5b0cd201dbb7214891652ea82f2b97880f4aa516

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala [hunk 5]
```diff
@@ -355,6 +385,33 @@ class RapidsShuffleClient(
     clientCopyExecutor: Executor,
     maximumMetadataSize: Long) extends Logging {
 
+  private[this] var devStorage: RapidsDeviceMemoryStore = GpuShuffleEnv.getDeviceStorage
+  private[this] var catalog: ShuffleReceivedBufferCatalog = GpuShuffleEnv.getReceivedCatalog
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:389]: Nit: I'd prefer to make these `val` and have the standard constructor default to these values rather than mutate them with an alternate constructor.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala [hunk 6]
```diff
@@ -593,10 +648,21 @@ class RapidsShuffleClient(
...
     connection.receive(buffersToReceive,
       tx => {
-        // TODO: during code review we agreed to make this receive per buffer, s.t. buffers could
-        //  begin go get freed earlier and likely out of order.
-        asyncOnCopyThread(HandleBounceBufferReceive(tx, bufferReceiveState,
-          currentRequest, buffersToReceive))
+        tx.getStatus match {
+          case TransactionStatus.Success =>
+            // TODO: during code review we agreed to make this receive per buffer, s.t.
+            //  buffers could begin go get freed earlier and likely out of order.
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:654]:
```suggestion
            //  buffers could be freed earlier and likely out of order.
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 7]
```diff
@@ -16,15 +16,14 @@
...
 package com.nvidia.spark.rapids.shuffle
 
-import java.util.concurrent.LinkedBlockingQueue
-
-import scala.collection.mutable
-
 import ai.rapids.cudf.{NvtxColor, NvtxRange}
 import com.nvidia.spark.rapids._
+import java.util.concurrent.LinkedBlockingQueue
+import scala.collection.mutable
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:22]: Why were these imports moved around?  Normally scala imports are separate per [the recommended style import layout.](https://github.com/NVIDIA/spark-rapids/blob/branch-0.2/docs/dev/idea-code-style-settings.md)

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 8]
```diff
@@ -47,14 +47,60 @@ class RapidsShuffleIterator(
...
+    taskContext: TaskContext,
     metricsUpdater: ShuffleMetricsUpdater)
   extends Iterator[ColumnarBatch]
     with Logging {
 
-  // Used for the producer/consumer pattern, where the producers are a
-  // set of [[RapidsShuffleClient]]s, and the consumer is the executor task,
-  // which is blocked in the iterator's next call
-  private[this] val resolvedBatches = new LinkedBlockingQueue[ShuffleReceivedBufferId]()
+  private[this] var catalog: ShuffleReceivedBufferCatalog = GpuShuffleEnv.getReceivedCatalog
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:55]: Same nit re: `val` instead of `var` that is mutated in alternate constructor.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 9]
```diff
@@ -127,24 +171,41 @@ class RapidsShuffleIterator(
             * about it.
             */
           blockId._1 match {
-            case sbbid: ShuffleBlockBatchId => sbbid
+            case sbbid: ShuffleBlockBatchId => (sbbid, blockId._3)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:174]: Nit: We should consider using a case class.  I find generic tuples beyond 2 to often be hard to follow without useful names for things, e.g.: needing to recall what the `Int` field really means.  A `case class` makes that obvious when reading the code.  Tuples not so much.  Doesn't have to be done as part of this PR.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 10]
```diff
@@ -194,21 +265,26 @@ class RapidsShuffleIterator(
...
+      resolvedBatches.forEach {
+        case BufferReceived(bufferId) =>
+          GpuShuffleEnv.getReceivedCatalog.removeBuffer(bufferId)
+        case _ =>
+      }
     }
   }
 
+  //TODO: on task completion we currently don't ask clients to stop/clean resources
+  taskContext.addTaskCompletionListener[Unit] { _ =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:281]: Rather than passing a task context explicitly for unit tests and needing to mock one up in the unit tests, it's simpler to do this instead:
```scala
Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => receiveBufferCleaner()))
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 11]
```diff
@@ -235,23 +311,35 @@ class RapidsShuffleIterator(
...
+          metricsUpdater.update(blockedTime, 1, sb.size, cb.numRows())
+        } finally {
+          nvtxRangeAfterGettingBatch.close()
+          range.close()
+          if (sb != null) {
+            sb.close()
+          }
+          catalog.removeBuffer(bufferId)
+        }
+      case err @ TransferError(blockManagerId, shuffleBlockBatchId, mapIndex, errorMessage) =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:332]: `err` isn't used, so why not just:
```scala
case TransferError(blockManagerId, shuffleBlockBatchId, mapIndex, errorMessage) =>
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManager.scala [hunk 12]
```diff
@@ -295,7 +305,7 @@ class RapidsShuffleInternalManager(conf: SparkConf, isDriver: Boolean)
       metrics: ShuffleReadMetricsReporter): ShuffleReader[K, C] = {
     // NOTE: This type of reader is not possible for gpu shuffle, as we'd need
     // to use the optimization within our manager, and we don't.
-    wrapped.getReaderForRange(unwrapHandle(handle), startMapIndex, endMapIndex,
+   wrapped.getReaderForRange(unwrapHandle(handle), startMapIndex, endMapIndex,
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManager.scala:308]: Indentation is off here

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala (hunk 5 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleClient.scala:389]: Nice, that's a lot better.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala (hunk 7 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:22]: Short answer, I hadn't applied the layout settings correctly, and somehow `mvn verify` is not picking this up, even with the `ImportOrderChecker` scalatyle rule.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala (hunk 9 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:174]: Sounds good.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleBufferCatalog.scala (hunk 2 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleBufferCatalog.scala:211]: So yes, the assumption is that the buffer is not being asked by anyone else, since it hasn't been advertised to the tracker of its existence. That said, it could be getting spilled at the same time, so I think there could be a race there.

The iterator has a very similar call. Where it will remove a received buffer from the `ShuffleReceivedBufferCatalog` when handing it off to the task (and the implementation is the same).

I am not sure at this stage is spilling is an issue here, do you have concerns with that case? I'll look more at that path too.

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShuffleBufferCatalog.scala:211]: I don't think spilling is an issue here.  `RapidsBufferCatalog` should already properly handle race conditions where a buffer mapping is being updated via spill just as it is being removed separately from what `ShuffleReceivedBufferCatalog` does.  A spill doesn't change shuffle block IDs or table IDs so `ShuffleReceivedBufferCatalog` doesn't know or care that spills are occuring -- that's an implementation detail hidden by `RapidsBufferCatalog`.

So if the synchronization isn't necessary because we fully expect no more than one thread to ever request a particular buffer, we can cover that with a comment here and/or at the top of the shuffle buffer catalog class.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala [hunk 13]
```diff
@@ -152,6 +132,8 @@ class RapidsShuffleIterator(
 
   private var started: Boolean = false
 
+  case class BlockIdMapIndex(id: ShuffleBlockBatchId, mapIndex: Int)
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:135]: Curious, how is `mapIndex` different than `id.mapId`?  If it's different a comment explaining would help.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/RapidsShuffleIterator.scala:135]: `mapId` used to be `mapIndex`, but it is now `taskAttemptId` as long as you are using the new shuffle protocol (default ON). `mapIndex` is the index into the `statuses` array in `MapOutputTracker`.

`reduceId` is the index into the `sizes` array, within a `MapStatus`.

This function summarizes it pretty well I think: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/MapOutputTracker.scala#L949.

I'll add a comment as suggested.

**Comment by abellina** [comment]
build
